File: packages/dtj-go/protocol.go
package dtj

import (
	"bytes"
	"crypto/rand"
	"encoding/binary"
	"fmt"
	"io"
	"math"
	"time"
)

// Protocol constants
const (
	ProtocolVersion = 1
	MaxFrameSize    = 1_048_576 // 1 MiB
)

// Opcodes
const (
	// Client -> Server
	OpHello         = 0x01
	OpOpenSession   = 0x02
	OpAppendEvent   = 0x03
	OpFinishSession = 0x04
	OpPing          = 0x05
	OpIntern        = 0x06

	// Server -> Client
	OpHelloOk         = 0x81
	OpOpenSessionOk   = 0x82
	OpAppendEventOk   = 0x83
	OpFinishSessionOk = 0x84
	OpPong            = 0x85
	OpInternOk        = 0x86
	OpError           = 0xFF
)

// Dictionary kinds
const (
	DictKindDomain    = 1
	DictKindCategory  = 2
	DictKindEventName = 3
	DictKindString    = 4
)

// Severity levels (match dtj::Severity)
type Severity uint8

const (
	SeverityDebug Severity = 0
	SeverityInfo  Severity = 1
	SeverityWarn  Severity = 2
	SeverityError Severity = 3
	SeverityFatal Severity = 4
)

// Type tags (match dtj::Value)
type TypeTag uint8

const (
	TypeTagBool     TypeTag = 0x01
	TypeTagI32      TypeTag = 0x02
	TypeTagI64      TypeTag = 0x03
	TypeTagU32      TypeTag = 0x04
	TypeTagU64      TypeTag = 0x05
	TypeTagF32      TypeTag = 0x06
	TypeTagF64      TypeTag = 0x07
	TypeTagEnum     TypeTag = 0x08
	TypeTagVec2F32  TypeTag = 0x09
	TypeTagVec3F32  TypeTag = 0x0A
	TypeTagInterned TypeTag = 0x0B
	TypeTagBytes    TypeTag = 0x0C
)

// Frame represents a protocol frame
type Frame struct {
	Opcode byte
	Body   []byte
}

// EncodeFrame encodes a frame with length prefix
func EncodeFrame(opcode byte, body []byte) ([]byte, error) {
	length := 1 + len(body) // includes opcode
	if length > MaxFrameSize {
		return nil, &ProtocolError{Opcode: opcode, Msg: fmt.Sprintf("frame too large: %d > %d", length, MaxFrameSize)}
	}
	buf := make([]byte, 4+length)
	binary.LittleEndian.PutUint32(buf[0:4], uint32(length))
	buf[4] = opcode
	copy(buf[5:], body)
	return buf, nil
}

// DecodeFrame decodes a frame from a reader
func DecodeFrame(r io.Reader) (*Frame, error) {
	// Read length (4 bytes)
	lenBuf := make([]byte, 4)
	if _, err := io.ReadFull(r, lenBuf); err != nil {
		return nil, &ProtocolError{Msg: fmt.Sprintf("failed to read frame length: %v", err)}
	}
	length := binary.LittleEndian.Uint32(lenBuf)
	if length > MaxFrameSize {
		return nil, &ProtocolError{Msg: fmt.Sprintf("frame too large: %d", length)}
	}

	// Read payload (length bytes including opcode)
	payload := make([]byte, length)
	if _, err := io.ReadFull(r, payload); err != nil {
		return nil, &ProtocolError{Msg: fmt.Sprintf("failed to read frame payload: %v", err)}
	}

	return &Frame{
		Opcode: payload[0],
		Body:   payload[1:],
	}, nil
}

// EncodeHello encodes a Hello frame
func EncodeHello() ([]byte, error) {
	body := make([]byte, 4)
	binary.LittleEndian.PutUint32(body, ProtocolVersion)
	return EncodeFrame(OpHello, body)
}

// DecodeHelloOk decodes a HelloOk response
func DecodeHelloOk(body []byte) (uint32, error) {
	if len(body) != 4 {
		return 0, &ProtocolError{Msg: "HelloOk body must be 4 bytes"}
	}
	return binary.LittleEndian.Uint32(body), nil
}

// OpenSessionMetadata holds metadata for OpenSession
type OpenSessionMetadata struct {
	FileName        string
	SessionID       [16]byte
	StartUtcUnixMs  int64
	MonoOriginNs    uint64
	ProducerName    string
	ProducerVersion string
}

// NewOpenSessionMetadata creates metadata with auto-generated values
func NewOpenSessionMetadata(fileName, producerName, producerVersion string) (*OpenSessionMetadata, error) {
	if len(producerName) > 32 {
		return nil, &ValueError{Msg: "producer_name must be <= 32 bytes"}
	}
	if len(producerVersion) > 16 {
		return nil, &ValueError{Msg: "producer_version must be <= 16 bytes"}
	}

	var sessionID [16]byte
	if _, err := rand.Read(sessionID[:]); err != nil {
		return nil, &ProtocolError{Msg: fmt.Sprintf("failed to generate session ID: %v", err)}
	}

	return &OpenSessionMetadata{
		FileName:        fileName,
		SessionID:       sessionID,
		StartUtcUnixMs:  time.Now().UnixMilli(),
		MonoOriginNs:    uint64(time.Now().UnixNano()), // Using wall clock as approximation
		ProducerName:    producerName,
		ProducerVersion: producerVersion,
	}, nil
}

// EncodeOpenSession encodes an OpenSession frame
func EncodeOpenSession(meta *OpenSessionMetadata) ([]byte, error) {
	fileNameBytes := []byte(meta.FileName)
	producerNameBytes := []byte(meta.ProducerName)
	producerVersionBytes := []byte(meta.ProducerVersion)

	body := new(bytes.Buffer)
	binary.Write(body, binary.LittleEndian, uint16(len(fileNameBytes)))
	body.Write(fileNameBytes)
	body.Write(meta.SessionID[:])
	binary.Write(body, binary.LittleEndian, meta.StartUtcUnixMs)
	binary.Write(body, binary.LittleEndian, meta.MonoOriginNs)
	binary.Write(body, binary.LittleEndian, uint16(len(producerNameBytes)))
	body.Write(producerNameBytes)
	binary.Write(body, binary.LittleEndian, uint16(len(producerVersionBytes)))
	body.Write(producerVersionBytes)

	return EncodeFrame(OpOpenSession, body.Bytes())
}

// EncodeIntern encodes an Intern request
func EncodeIntern(kind uint8, name string) ([]byte, error) {
	nameBytes := []byte(name)
	if len(nameBytes) > 1024 {
		return nil, &ValueError{Msg: "name too long (max 1024 bytes)"}
	}

	body := new(bytes.Buffer)
	body.WriteByte(kind)
	binary.Write(body, binary.LittleEndian, uint16(len(nameBytes)))
	body.Write(nameBytes)

	return EncodeFrame(OpIntern, body.Bytes())
}

// DecodeInternOk decodes an InternOk response
func DecodeInternOk(body []byte) (uint32, error) {
	if len(body) != 4 {
		return 0, &ProtocolError{Msg: "InternOk body must be 4 bytes"}
	}
	return binary.LittleEndian.Uint32(body), nil
}

// EncodeAppendEvent encodes an AppendEvent frame with single field (MVP)
func EncodeAppendEvent(
	monotonicNs uint64,
	domainID, categoryID, eventNameID, correlationID uint32,
	severity Severity,
	fieldNameID uint32,
	typeTag TypeTag,
	valueBody []byte,
) ([]byte, error) {
	body := new(bytes.Buffer)
	binary.Write(body, binary.LittleEndian, monotonicNs)
	binary.Write(body, binary.LittleEndian, domainID)
	binary.Write(body, binary.LittleEndian, categoryID)
	binary.Write(body, binary.LittleEndian, eventNameID)
	binary.Write(body, binary.LittleEndian, correlationID)
	body.WriteByte(byte(severity))
	binary.Write(body, binary.LittleEndian, uint16(1)) // field_count = 1
	binary.Write(body, binary.LittleEndian, fieldNameID)
	body.WriteByte(byte(typeTag))
	body.Write([]byte{0, 0, 0}) // reserved
	body.Write(valueBody)

	return EncodeFrame(OpAppendEvent, body.Bytes())
}

// DecodeAppendEventOk decodes an AppendEventOk response
func DecodeAppendEventOk(body []byte) (uint64, error) {
	if len(body) != 8 {
		return 0, &ProtocolError{Msg: "AppendEventOk body must be 8 bytes"}
	}
	return binary.LittleEndian.Uint64(body), nil
}

// EncodeFinishSession encodes a FinishSession frame
func EncodeFinishSession() ([]byte, error) {
	return EncodeFrame(OpFinishSession, nil)
}

// EncodePing encodes a Ping frame
func EncodePing() ([]byte, error) {
	return EncodeFrame(OpPing, nil)
}

// DecodeError decodes an Error frame
func DecodeError(body []byte) string {
	return string(body)
}

// EncodeValue encodes a Go value to (typeTag, valueBody)
func EncodeValue(v any) (TypeTag, []byte, error) {
	switch val := v.(type) {
	case bool:
		b := byte(0)
		if val {
			b = 1
		}
		return TypeTagBool, []byte{b}, nil

	case int8:
		return TypeTagI64, int64ToBytes(int64(val)), nil
	case int16:
		return TypeTagI64, int64ToBytes(int64(val)), nil
	case int32:
		return TypeTagI64, int64ToBytes(int64(val)), nil
	case int64:
		return TypeTagI64, int64ToBytes(val), nil
	case int:
		// Check if fits in int64
		if val >= math.MinInt64 && val <= math.MaxInt64 {
			return TypeTagI64, int64ToBytes(int64(val)), nil
		}
		return 0, nil, &ValueError{Msg: fmt.Sprintf("int out of range: %d", val)}

	case uint8:
		return TypeTagU64, uint64ToBytes(uint64(val)), nil
	case uint16:
		return TypeTagU64, uint64ToBytes(uint64(val)), nil
	case uint32:
		return TypeTagU64, uint64ToBytes(uint64(val)), nil
	case uint64:
		return TypeTagU64, uint64ToBytes(val), nil
	case uint:
		return TypeTagU64, uint64ToBytes(uint64(val)), nil

	case float32:
		return TypeTagF64, float64ToBytes(float64(val)), nil
	case float64:
		return TypeTagF64, float64ToBytes(val), nil

	case string:
		// Strings must be interned separately
		return 0, nil, &ValueError{Msg: "strings must be interned first; use field_name_id from intern"}

	case []byte:
		if len(val) > math.MaxUint32 {
			return 0, nil, &ValueError{Msg: "bytes too long"}
		}
		body := make([]byte, 4+len(val))
		binary.LittleEndian.PutUint32(body[0:4], uint32(len(val)))
		copy(body[4:], val)
		return TypeTagBytes, body, nil

	default:
		return 0, nil, &ValueError{Msg: fmt.Sprintf("unsupported value type: %T", v)}
	}
}

func int64ToBytes(v int64) []byte {
	b := make([]byte, 8)
	binary.LittleEndian.PutUint64(b, uint64(v))
	return b
}

func uint64ToBytes(v uint64) []byte {
	b := make([]byte, 8)
	binary.LittleEndian.PutUint64(b, v)
	return b
}

func float64ToBytes(v float64) []byte {
	b := make([]byte, 8)
	binary.LittleEndian.PutUint64(b, math.Float64bits(v))
	return b
}

// SeverityFromString converts a string to Severity
func SeverityFromString(s string) (Severity, error) {
	switch s {
	case "debug":
		return SeverityDebug, nil
	case "info":
		return SeverityInfo, nil
	case "warn":
		return SeverityWarn, nil
	case "error":
		return SeverityError, nil
	case "fatal":
		return SeverityFatal, nil
	default:
		return 0, &ValueError{Msg: fmt.Sprintf("invalid severity: %s", s)}
	}
}

// String returns the string representation of Severity
func (s Severity) String() string {
	switch s {
	case SeverityDebug:
		return "debug"
	case SeverityInfo:
		return "info"
	case SeverityWarn:
		return "warn"
	case SeverityError:
		return "error"
	case SeverityFatal:
		return "fatal"
	default:
		return "unknown"
	}
}

File: packages/dtj-rust/src/protocol.rs
use crate::error::Error;
use std::io::{Read, Write};

// =============================================================================
// Frame read/write
// =============================================================================

/// Write a frame: 4-byte little-endian length (including opcode) + opcode + payload
pub fn write_frame<W: Write>(writer: &mut W, opcode: u8, payload: &[u8]) -> Result<(), Error> {
    let len = 1u32 + payload.len() as u32;
    let mut frame = Vec::with_capacity(4 + 1 + payload.len());
    frame.extend_from_slice(&len.to_le_bytes());
    frame.push(opcode);
    frame.extend_from_slice(payload);
    writer.write_all(&frame).map_err(|_| Error::IoError)?;
    Ok(())
}

/// Read a frame: 4-byte length + opcode + payload
pub fn read_frame<R: Read>(reader: &mut R) -> Result<(u8, Vec<u8>), Error> {
    let mut len_buf = [0u8; 4];
    reader
        .read_exact(&mut len_buf)
        .map_err(|_| Error::IoError)?;
    let len = u32::from_le_bytes(len_buf) as usize;
    if len == 0 || len > 1024 * 1024 {
        return Err(Error::FrameTooLarge);
    }
    let mut frame = vec![0u8; len];
    reader.read_exact(&mut frame).map_err(|_| Error::IoError)?;
    let opcode = frame[0];
    let payload = frame[1..].to_vec();
    Ok((opcode, payload))
}

// =============================================================================
// Hello / HelloOk
// =============================================================================

pub const OPCODE_HELLO: u8 = 0x01;
pub const OPCODE_HELLO_OK: u8 = 0x81;
pub const PROTOCOL_VERSION: u32 = 1;

pub fn write_hello<W: Write>(writer: &mut W) -> Result<(), Error> {
    write_frame(writer, OPCODE_HELLO, &PROTOCOL_VERSION.to_le_bytes())
}

pub fn read_hello_ok<R: Read>(reader: &mut R) -> Result<(), Error> {
    let (opcode, payload) = read_frame(reader)?;
    if opcode != OPCODE_HELLO_OK {
        return Err(Error::Protocol);
    }
    if payload.len() != 4 {
        return Err(Error::Protocol);
    }
    let version = u32::from_le_bytes(payload[..4].try_into().unwrap());
    if version != PROTOCOL_VERSION {
        return Err(Error::BadVersion);
    }
    Ok(())
}

/// Read HelloOk or Error frame.
/// Returns Ok(true) for HelloOk, Ok(false) for Error, Err for other frames.
pub fn read_hello_ok_or_error<R: Read>(reader: &mut R) -> Result<bool, Error> {
    let (opcode, _payload) = read_frame(reader)?;
    match opcode {
        OPCODE_HELLO_OK => Ok(true),
        OPCODE_ERROR => Ok(false),
        _ => Err(Error::Protocol),
    }
}

// =============================================================================
// OpenSession / OpenSessionOk
// =============================================================================

pub const OPCODE_OPEN_SESSION: u8 = 0x02;
pub const OPCODE_OPEN_SESSION_OK: u8 = 0x82;

/// Payload for OpenSession frame
#[derive(Debug, Clone)]
pub struct OpenSessionPayload {
    pub file_name: String,
    pub session_id: [u8; 16],
    pub start_utc_unix_ms: i64,
    pub mono_origin_ns: u64,
    pub producer_name: String,
    pub producer_version: String,
}

pub fn write_open_session<W: Write>(writer: &mut W, p: &OpenSessionPayload) -> Result<(), Error> {
    let mut buf = Vec::new();
    let file_name = p.file_name.as_bytes();
    buf.extend_from_slice(&(file_name.len() as u16).to_le_bytes());
    buf.extend_from_slice(file_name);
    buf.extend_from_slice(&p.session_id);
    buf.extend_from_slice(&p.start_utc_unix_ms.to_le_bytes());
    buf.extend_from_slice(&p.mono_origin_ns.to_le_bytes());
    let producer_name = p.producer_name.as_bytes();
    buf.extend_from_slice(&(producer_name.len() as u16).to_le_bytes());
    buf.extend_from_slice(producer_name);
    let producer_version = p.producer_version.as_bytes();
    buf.extend_from_slice(&(producer_version.len() as u16).to_le_bytes());
    buf.extend_from_slice(producer_version);
    write_frame(writer, OPCODE_OPEN_SESSION, &buf)
}

pub fn read_open_session_ok<R: Read>(reader: &mut R) -> Result<(), Error> {
    let (opcode, payload) = read_frame(reader)?;
    if opcode != OPCODE_OPEN_SESSION_OK {
        return Err(Error::Protocol);
    }
    if !payload.is_empty() {
        return Err(Error::Protocol);
    }
    Ok(())
}

/// Read OpenSessionOk or Error frame.
/// Returns Ok(true) for OpenSessionOk, Ok(false) for Error, Err for other frames.
pub fn read_open_session_ok_or_error<R: Read>(reader: &mut R) -> Result<bool, Error> {
    let (opcode, payload) = read_frame(reader)?;
    match opcode {
        OPCODE_OPEN_SESSION_OK if payload.is_empty() => Ok(true),
        OPCODE_ERROR => Ok(false),
        _ => Err(Error::Protocol),
    }
}

// =============================================================================
// Intern / InternOk
// =============================================================================

pub const OPCODE_INTERN: u8 = 0x06;
pub const OPCODE_INTERN_OK: u8 = 0x86;

pub const DICT_KIND_DOMAIN: u8 = 1;
pub const DICT_KIND_CATEGORY: u8 = 2;
pub const DICT_KIND_EVENT_NAME: u8 = 3;
pub const DICT_KIND_STRING: u8 = 4;

pub fn write_intern<W: Write>(writer: &mut W, dict_kind: u8, name: &str) -> Result<(), Error> {
    let name_bytes = name.as_bytes();
    let mut payload = Vec::with_capacity(1 + 2 + name_bytes.len());
    payload.push(dict_kind);
    payload.extend_from_slice(&(name_bytes.len() as u16).to_le_bytes());
    payload.extend_from_slice(name_bytes);
    write_frame(writer, OPCODE_INTERN, &payload)
}

pub fn read_intern_ok<R: Read>(reader: &mut R) -> Result<u32, Error> {
    let (opcode, payload) = read_frame(reader)?;
    if opcode != OPCODE_INTERN_OK {
        return Err(Error::Protocol);
    }
    if payload.len() != 4 {
        return Err(Error::Protocol);
    }
    Ok(u32::from_le_bytes(payload[..4].try_into().unwrap()))
}

// =============================================================================
// AppendEvent / AppendEventOk
// =============================================================================

pub const OPCODE_APPEND_EVENT: u8 = 0x03;
pub const OPCODE_APPEND_EVENT_OK: u8 = 0x83;

/// Frame for writing an AppendEvent record.
pub struct AppendEventFrame<'a> {
    pub monotonic_ns: u64,
    pub domain_id: u32,
    pub category_id: u32,
    pub event_name_id: u32,
    pub correlation_id: u32,
    pub severity: u8,
    pub field_name_id: u32,
    pub type_tag: u8,
    pub value_body: &'a [u8],
}

pub fn write_append_event<W: Write>(
    writer: &mut W,
    frame: AppendEventFrame<'_>,
) -> Result<(), Error> {
    let mut payload = Vec::new();
    payload.extend_from_slice(&frame.monotonic_ns.to_le_bytes());
    payload.extend_from_slice(&frame.domain_id.to_le_bytes());
    payload.extend_from_slice(&frame.category_id.to_le_bytes());
    payload.extend_from_slice(&frame.event_name_id.to_le_bytes());
    payload.extend_from_slice(&frame.correlation_id.to_le_bytes());
    payload.push(frame.severity);
    payload.extend_from_slice(&1u16.to_le_bytes()); // field_count = 1
    payload.extend_from_slice(&frame.field_name_id.to_le_bytes());
    payload.push(frame.type_tag);
    payload.extend_from_slice(&[0u8; 3]); // reserved
    payload.extend_from_slice(frame.value_body);
    write_frame(writer, OPCODE_APPEND_EVENT, &payload)
}

pub fn read_append_event_ok<R: Read>(reader: &mut R) -> Result<u64, Error> {
    let (opcode, payload) = read_frame(reader)?;
    if opcode != OPCODE_APPEND_EVENT_OK {
        return Err(Error::Protocol);
    }
    if payload.len() != 8 {
        return Err(Error::Protocol);
    }
    Ok(u64::from_le_bytes(payload[..8].try_into().unwrap()))
}

// =============================================================================
// FinishSession / FinishSessionOk
// =============================================================================

pub const OPCODE_FINISH_SESSION: u8 = 0x04;
pub const OPCODE_FINISH_SESSION_OK: u8 = 0x84;

pub fn write_finish_session<W: Write>(writer: &mut W) -> Result<(), Error> {
    write_frame(writer, OPCODE_FINISH_SESSION, &[])
}

pub fn read_finish_session_ok<R: Read>(reader: &mut R) -> Result<(), Error> {
    let (opcode, payload) = read_frame(reader)?;
    if opcode != OPCODE_FINISH_SESSION_OK {
        return Err(Error::Protocol);
    }
    if !payload.is_empty() {
        return Err(Error::Protocol);
    }
    Ok(())
}

/// Read FinishSessionOk or Error frame.
/// Returns Ok(true) for FinishSessionOk, Ok(false) for Error, Err for other frames.
pub fn read_finish_session_ok_or_error<R: Read>(reader: &mut R) -> Result<bool, Error> {
    let (opcode, payload) = read_frame(reader)?;
    match opcode {
        OPCODE_FINISH_SESSION_OK if payload.is_empty() => Ok(true),
        OPCODE_ERROR => Ok(false),
        _ => Err(Error::Protocol),
    }
}

// =============================================================================
// Error frame
// =============================================================================

pub const OPCODE_ERROR: u8 = 0xFF;

pub fn read_error_frame<R: Read>(reader: &mut R) -> Result<String, Error> {
    let (opcode, payload) = read_frame(reader)?;
    if opcode != OPCODE_ERROR {
        return Err(Error::Protocol);
    }
    String::from_utf8(payload).map_err(|_| Error::Protocol)
}
